File: subsystem_cpp/subsystem/sdk/capplication.cpp
#include "capplication.h"
#include "chost_p.h"

struct _self_CApplication {
    CApplicationDelegate::ref delegate;
};

static void onAppCreate(shared_ptr<_self_CApplication> self) {
    if (self->delegate != nullptr) {
        self->delegate->applicationDidFinishLaunching();
    }
}

static void onUIAppear(shared_ptr<_self_CApplication> self) {
    if (self->delegate != nullptr) {
        self->delegate->applicationDidBecomeActive();
    }
}

static void onUIDisappear(shared_ptr<_self_CApplication> self) {
    if (self->delegate != nullptr) {
        self->delegate->applicationDidEnterBackground();
    }
}

CApplication::ref CApplication::sharedApplication() {
    static CApplication::ref object;
    if (object == nullptr) {
        object = CApplication::create();
    }
    return object;
}

CApplication::CApplication() {
    C_INIT(self);
    
    CHostAddListener(CHostEventAppCreate  , bind(onAppCreate  , self));
    CHostAddListener(CHostEventUIAppear   , bind(onUIAppear   , self));
    CHostAddListener(CHostEventUIDisappear, bind(onUIDisappear, self));
}

void CApplication::setDelegate(CApplicationDelegate::ref delegate) {
    self->delegate = delegate;
}

CApplicationDelegate::ref CApplication::delegate() {
    return self->delegate;
}

void CApplicationMain(CApplicationDelegate::ref delegate) {
    CApplication::sharedApplication()->setDelegate(delegate);
}

File: subsystem_lua/subsystem/sdk/capplication.lua
CApplication = extends(CObject, {
    theObject = nil,
    delegate = nil,
})

CApplication.Delegate = extends(CObject, {
    applicationDidFinishLaunching = function() end,
    applicationDidBecomeActive    = function() end,
    applicationDidEnterBackground = function() end,
})

function CApplication:sharedApplication()
    if CApplication.theObject == nil then
        CApplication.theObject = CApplication:new()
    end
    return CApplication.theObject
end

local function onAppCreate(self)
    if self.delegate ~= nil then
        self.delegate:applicationDidFinishLaunching()
    end
end

local function onUIAppear(self)
    if self.delegate ~= nil then
        self.delegate:applicationDidBecomeActive()
    end
end

local function onUIDisappear(self)
    if self.delegate ~= nil then
        self.delegate:applicationDidEnterBackground()
    end
end

function CApplication:init()
    CApplication.super.init(self)

    local a = self
    CHost:addListener(CHost.Event.AppCreate  , function() onAppCreate  (a) end)
    CHost:addListener(CHost.Event.UIAppear   , function() onUIAppear   (a) end)
    CHost:addListener(CHost.Event.UIDisappear, function() onUIDisappear(a) end)
end

function CApplication:setDelegate(delegate)
    self.delegate = delegate
end

function CApplication:getDelegate()
    return self.delegate
end

function CApplicationMain(delegate)
    CApplication:sharedApplication():setDelegate(delegate)
end
